[RDY]:hi Normal works with -u NONE

teto opened the pull request:
The PR has 2 commits, the first one makes `:hi Normal works with -u NONE` works, just like vim.
This first commit also:
 - Makes HL_TABLE and ATTR_ENTTRY a function instead of a macro so that in can be used in gdb.
 - Introduces ATTRENTRY_INIT to init attrentry_tchanges, they make sense in the context of https://github.com/neovim/neovim/pull/6566 ; the long term goal is to be able to distinguish between NONE and an unset color.

The 2nd commit is just some documentation/styling changes (brackets) I had made to fix linting issues in another PR. I salvaged them but they are not necessary (might need to do another pass at it) so I can ditch it or send it another PR it that.

Nb: https://github.com/neovim/neovim/pull/6566 had become a mess so I am trying to split it into meaningful commits.

jamessan commented on src/nvim/syntax.c:
```diff
@@ -7186,20 +7215,27 @@ int syn_namen2id(char_u *linep, int len)
+
+  // HACK we need to make sure that the first highlight is the "Normal" hl
+  if (highlight_ga.ga_len == 0) {
+    int res = syn_add_group((char_u *)"Normal");
```
`syn_add_group()` requires its argument to be an allocated string.  You're passing a static string instead.  Try this:

```c
int res = syn_add_group((char_u *)xstrdup("Normal"));
```

jamessan commented on src/nvim/syntax_defs.h:
```diff
@@ -73,4 +73,14 @@ typedef struct attr_entry {
   int cterm_fg_color, cterm_bg_color;
 } attrentry_T;
 
+#define ATTRENTRY_INIT {.rgb_ae_attr = 0, .cterm_ae_attr = 0, \
```
Why not put `rgb_ae_attr` and `cterm_ae_attr` on their own lines too?  It looks a little odd having these two here but everything else on its own line.

teto replied:
done

jamessan commented on src/nvim/syntax.c:
```diff
@@ -7186,20 +7217,27 @@ int syn_namen2id(char_u *linep, int len)
+  // HACK we need to make sure that the first highlight is the "Normal" hl
+  if (highlight_ga.ga_len == 0) {
+    int res = syn_add_group((char_u *)xstrdup("Normal"));
+    assert(res == 0);
```
Need to add a `(void)res;` line so release builds think `res` is used.

teto replied:
Indeed. I would like to move the Normal hl creation to init_highlight but that would need more thoughts/tests as I am afraid ` :high Normal NONE`  or some other command could break things.

jamessan commented:
```
From test_signs.vim:
Found errors in Test_sign():
function RunTheTest[9]..Test_sign line 19: Pattern '^sign Sign1 text=x sign Sign2 icon=../../pixmaps/stock_vim_find_help.png .*text=xy linehl=Error texthl=Title$' does not match 'sign Sign1 text=x sign Sign2 icon=../../pixmaps/stock_vim_find_help.png (not supported) text=xy linehl=MatchParen texthl=VertSplit'
```

teto commented:
not sure https://travis-ci.org/neovim/neovim/jobs/251093995#L2732 is related  to my changes:
```
 [----------] Running tests from /home/travis/build/neovim/neovim/test/functional/autocmd/termclose_spec.lua

[ RUN      ] TermClose event triggers when terminal job ends: -- Output to stderr:

Vim: Caught deadly signal 'SIGHUP'
```
is it possible to restart the failed tests please ?

jamessan commented:
> is it possible to restart the failed tests please ?

Done.

teto committed:
syntax.c: register 'Normal' highlight group

- :hi Normal works with -u NONE
- Makes HL_TABLE and ATTR_ENTYRY a function instead of a macro so that in can be used in gdb.
- Introduces ATTRENTRY_INIT to init attrentry_t

teto committed:
syntax.c: style

Converts some documentation to doxygen format + minor styling
improvements.

teto commented:
I updated the first post to better reflect the current content.

teto commented:
Would be happy to see this merged (even if only the first commit) as it solves the `:hi Normal`problem and helps me fixing https://github.com/neovim/neovim/pull/7082

justinmk commented on src/nvim/syntax.c:
```diff
@@ -6270,12 +6270,12 @@ do_highlight(
 
-  if (STRCMP(HL_TABLE()[idx].sg_name_u, "NORMAL") == 0)
-    is_normal_group = TRUE;
+  is_normal_group = (STRCMP(HL_TABLE()[idx].sg_name_u, "NORMAL") == 0);
```
before, `is_normal_group` kept its value if the if-block was not entered. Is this change intentional?

teto replied:
it's the only place that sets is_normal_group so it's equivalent.

justinmk commented on src/nvim/syntax.c:
```diff
@@ -6707,14 +6711,16 @@ static void highlight_clear(int idx)
+/// Table with the specifications for an attribute number.
+/// Note that this table is used by ALL buffers.  This is required because the
+/// GUI can redraw at any time for any buffer.
+/// The attributes table
```
the last line seems a bit unnecessary

justinmk commented on src/nvim/syntax.c:
```diff
@@ -7124,10 +7143,10 @@ set_hl_attr (
- * Lookup a highlight group name and return it's ID.
- * If it is not found, 0 is returned.
- */
+/// Lookup a highlight group name and return it's ID.
```
"its"

justinmk commented on src/nvim/syntax.c:
```diff
@@ -7124,10 +7143,10 @@ set_hl_attr (
+/// Lookup a highlight group name and return it's ID.
+///
+/// @param highlight name e.g. 'Cursor', 'Normal'
+/// @return the highlight id, else 0 if it \p name does not exist
```
omit "it"

justinmk commented on src/nvim/syntax.c:
```diff
@@ -7195,11 +7214,11 @@ int syn_check_group(char_u *pp, int len)
- * "name" must be an allocated string, it will be consumed.
- * Return 0 for failure.
- */
+/// Add new highlight group and return it's ID.
```
"its"

justinmk commented:
Great to see smaller PRs, it helps a lot :>

LGTM, one comment: https://github.com/neovim/neovim/pull/6973#discussion_r134349010

teto commented:
The failure does not seem related to the changes:
```
[----------] Running tests from /home/travis/build/neovim/neovim/test/functional/autocmd/termclose_spec.lua

[ RUN      ] TermClose event triggers when fast-exiting terminal job stops: -- Output to stderr:

Vim: Caught deadly signal 'SIGHUP'

Vim: Finished.

CMake Error at /home/travis/build/neovim/neovim/cmake/RunTests.cmake:53 (message):

  Running functional tests failed with error: 1.
```

teto commented:
Thanks for the review, I rebased and pushed your fixes. Should be ok now.
